commit 346ad881f827d6a35b3e5da3d70ef7b05748d2c4 (refs/changes/88/2588/6)
Author: Orlando Vazquez <ovazquez@gmail.com>
Date:   2017-09-19T05:20:10+00:00 (2 years, 1 month ago)
    
    DOCKER-1106 Docker connection watcher functionality is causing user operations to fail when VMAPI is slow to respond to ping requests

diff --git a/lib/common.js b/lib/common.js
index 4a24015..da6f812 100644
--- a/lib/common.js
+++ b/lib/common.js
@@ -400,12 +400,16 @@ function uncaughtHandler(req, res, route, err) {
  * any request
  */
 function checkServices(req, res, next) {
-    req.app.connWatcher.checkAvailability(function (err) {
-        if (err) {
-            return next(err);
-        }
+    if (req.app.config.useConnectionStatusWatcher) {
+        req.app.connWatcher.checkAvailability(function (err) {
+            if (err) {
+                return next(err);
+            }
+            next();
+        });
+    } else {
         next();
-    });
+    }
 }
 
 
diff --git a/lib/docker.js b/lib/docker.js
index d1d05f7..2ee0fd0 100644
--- a/lib/docker.js
+++ b/lib/docker.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2016, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 /*
@@ -73,7 +73,11 @@ function App(opts) {
     var Backend = require('./backends/' + self.config.backend);
     self.backend = new Backend({log: self.log, config: self.config});
 
-    self.setupConnections();
+    if (self.config.useConnectionStatusWatcher) {
+        self.setupConnectionsWatcher();
+    } else {
+        self.setupConnections();
+    }
 
     // TODO make the other clients accessible via req.app
 
@@ -152,9 +156,33 @@ function App(opts) {
     self.setupAdminSever();
 }
 
+
 App.prototype.setupConnections = function setupConnections() {
     var self = this;
 
+    self.moray = self.createMorayClient();
+    self.ufds = self.createUfdsClient(self.config.ufds);
+    self.wfapi =  self.createWorkflowClient(self.config.wfapi);
+
+    self.cnapi = new CNAPI(self.config.cnapi);
+    self.vmapi = new VMAPI(self.config.vmapi);
+    self.imgapi = new IMGAPI(self.config.imgapi);
+};
+
+
+App.prototype.createWorkflowClient = function createWorkflowClient() {
+    var self = this;
+    var wfclient = new wfapi(self.config.wfapi, self.log);
+    wfclient.connect(function () {
+        self.log.info('wfapi is ready');
+    });
+    return wfclient;
+};
+
+
+App.prototype.setupConnectionsWatcher = function setupConnectionsWatcher() {
+    var self = this;
+
     self.connWatcher = new ConnectionStatusWatcher({
         app: self
     });
@@ -240,13 +268,7 @@ App.prototype.setupConnections = function setupConnections() {
         }
     });
 
-    self.createUfdsClient(self.config.ufds, function (err, ufds) {
-        if (err) {
-            self.log.error({ err: err }, 'ufds error');
-            return;
-        }
-        self.ufds = ufds;
-    });
+    self.ufds = self.createUfdsClient(self.config.ufds);
 };
 
 
@@ -426,12 +448,13 @@ App.prototype.createMorayClient = function createMorayClient() {
     return client;
 };
 
+
 /**
  * Creates a UFDS client instance pointing to the UFDS server provided
- * in options. callback will be called either with Error - cb(err) - or
- * with the recently instantiated client object: cb(null, ufds_client)
+ * in options.
  */
-App.prototype.createUfdsClient = function (options, callback) {
+
+App.prototype.createUfdsClient = function (options) {
     options.log = this.log;
     var ufds = new UFDS(options);
 
@@ -443,15 +466,15 @@ App.prototype.createUfdsClient = function (options, callback) {
         ufds.on('connect', function () {
             options.log.info('UFDS reconnected');
         });
-        callback(null, ufds);
     });
 
     ufds.once('error', function (err) {
         // You are screwed. It's likely that the bind credentials were bad.
         // Treat this as fatal and move on:
         options.log.error({err: err}, 'UFDS connection error');
-        callback(err);
     });
+
+    return ufds;
 };
 
 /*
diff --git a/lib/endpoints/build.js b/lib/endpoints/build.js
index a5ad4b2..4b0f9f6 100644
--- a/lib/endpoints/build.js
+++ b/lib/endpoints/build.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2014, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 var path = require('path');
@@ -93,7 +93,7 @@ function build(req, res, next) {
         req: req,
         req_id: req.getId(),
         res: res,
-        wfapi: req.wfapi
+        wfapi: req.app.wfapi
     }, function (err) {
         if (err) {
             log.error('docker build error', err);
diff --git a/sapi_manifests/docker/template b/sapi_manifests/docker/template
index 652cf95..ce04a8e 100644
--- a/sapi_manifests/docker/template
+++ b/sapi_manifests/docker/template
@@ -62,6 +62,9 @@
     "dockerRegistryInsecure": {{^docker_registry_insecure}}false{{/docker_registry_insecure}}{{#docker_registry_insecure}}{{{docker_registry_insecure}}}{{/docker_registry_insecure}},
 
     "useTls": {{^USE_TLS}}false{{/USE_TLS}}{{#USE_TLS}}{{{USE_TLS}}}{{/USE_TLS}},
+    {{#DOCKER_USE_CONN_WATCHER}}
+    "useConnectionStatusWatcher": true,
+    {{/DOCKER_USE_CONN_WATCHER}}
     "tls": {
         "requestCert": true,
         "rejectUnauthorized": false
